### ISSUE TITLE
Avoid re-running complex generators when checking many properties of a single operation

### ISSUE THREAD

**Issue by doughsay** [whatyouhide/stream_data#122]
So I've run into what I think is a bit of a usability issue with complex generators or slow operations, and I wanted to get some feedback.  Either there's something here, or I'm just abusing/misusing the library in some way, but please see the below code snippet to get an idea of what I'm doing:

```elixir
# This is slow: it regenerates complex_data_structure and runs slow_operation many times

property "property 1 holds for slow_operation" do
  check all complex_data_structure <- gen_data() do
    output = slow_operation(complex_data_structure)
    assert_property1(output)
  end
end

property "property 2 holds for slow_operation" do
  check all complex_data_structure <- gen_data() do
    output = slow_operation(complex_data_structure)
    assert_property2(output)
  end
end

property "property 3 holds for slow_operation" do
  check all complex_data_structure <- gen_data() do
    output = slow_operation(complex_data_structure)
    assert_property3(output)
  end
end

# This is only considered one property; failure in one assert causes the whole property to be considered failing

property "some properties holds for slow_operation" do
  check all complex_data_structure <- gen_data() do
    output = slow_operation(complex_data_structure)

    assert_property1(output)
    assert_property2(output)
    assert_property3(output)
  end
end

# What I think I want is something like this:

check all complex_data_structure <- gen_data() do
  output = slow_operation(complex_data_structure)

  property "property 1 holds for slow_operation" do
    assert_property1(output)
  end

  property "property 2 holds for slow_operation" do
    assert_property2(output)
  end

  property "property 3 holds for slow_operation" do
    assert_property3(output)
  end
end
```

Does the above make sense to want, or am I maybe going about things slightly wrong?

**Comment by whatyouhide** [comment]
I might be missing the point here. You want to have three separate properties for one generated data structure because you want to have two passing properties if one fails? If I understand this correctly, this is the same as when you have multiple tests that share a time-consuming setup: either you live with that, or you shove the tests together in one test so you run the setup only once, right?

I am trying to understand how this is different in properties compared to normal tests, so that we can get an idea if this is a stream_data problem specifically.

**Comment by whatyouhide** [comment]
@doughsay I don't feel like this is a stream_data specific problem, so I don't think we should try to solve this in stream_data. But thanks for reporting! 💟

**Comment by CatsOnFilm** [comment]
I would rephrase the issue as a feature request rather than a problem. I was looking for clarification also.
The problem described above is that the setup needs to be duplicated because there isn't an abstraction akin to `setup` or `setup_all` analogous to the prop-testing ideology. The property macro feels akin to the "test" macro of ExUnit (i.e. a way of making single assertions and reporting with a focused label). But when the setup is repeated from `property` to `property` call, it feels like there should be one/two more abstractions to isolate the setup to test against multiple `property` calls.

**Comment by whatyouhide** [comment]
@CatsOnFilm there's no `setup` equivalent because there is `setup` itself, right? If you want to define generators in `setup` you can do that and then use them in tests. But it's gonna be as expensive as defining them in each test. Please let me know if I'm missing something here.
